docs/developer/architecture/ci.rst

```diff
@@ -0,0 +1,122 @@
+CI
+==
+
+Pre-merge
+---------
+
+What should be tested in pre-merge on every ``user/*``, ``feature/*``,
+``improvement/*``, ``bugfix/*``, ``w/*`` branches. The pre-merge test should
+not long too much time (less than 40 minutes is great) so we can't
+test everything in pre-merge, we should only test building of the product
+and check that product still usable.
+
+- Building tests
+
+  - Build
+  - Lint
+  - Unit tests
+
+- Installation tests
+
+  - Simple install RHEL
+  - Simple install CentOs + expansion
+
+When merging several pull requests at the same time, given that we are on a
+queue branch (``q/*``), we may want to have or we may require additional tests
+as combinaison of several PRs could have larger impact than all individual PR.
+
+- Simple upgrade/downgrade
+
+.. _post-merge:
+
+Post-merge
+----------
+
+On each and every ``development/2.*`` branches we want to run complex tests
+to ensure that the current product continues to work.
+
+Nightly
+^^^^^^^
+
+- Upgrade, downgrade tests:
+
+  - For development branch just before
+
+    e.g.: on ``development/2.x`` test upgrade from ``development/2.(x-1)``
+    and downgrade to ``development/2.(x-1)``
+
+    - Build branch ``development/2.(x-1)``
+    - Tests:
+
+      - Single node test
+      - Complex deployment test
+
+  - For last released version of current minor
+
+    e.g.: on ``development/2.x`` when developing "2.x.y" test upgrade from
+    ``metalk8s-2.x.(y-1)`` and downgrade to ``metalk8s-2.x.(y-1)``
+
+    - Single node test
+    - Complex deployment test
+
+  - For last released version of minor just before
+
+    e.g.: on ``development/2.x`` when developing "2.x.y" test upgrade from
+    ``metalk8s-2.(x-1).z`` and downgrade to ``metalk8s-2.(x-1).z`` where
+    "2.(x-1).z" is the last patch released version for "2.(x-1)" (``z`` may be
+    different from ``y``)
+
+    - Single node test
+    - Complex deployment test
+
+- Backup, restore tests
+- Solutions tests
+
+.. todo::
+
+   - Define what is "complex deployment"
+
+Weekly
+^^^^^^
+
+More complex tests (TBD)
+
+Adaptive test plan
+------------------
+
+CI pre-merge may be more flexible by including some logic about the content
+of the changeset.
+
+For example:
+
+- For a commit that changes solely/uniquely documentation, we don't need/want
+  to run all the installation test but rather run tests related to
+  documentation.
+- For a commit touching upgrade orchestrate we want to test upgrade directly
+  in pre-merge and not wait :ref:`Post merge<post-merge>` build to get the
+  test result.
```

Comment:
Those are actual requirements, formulated through examples. Please abstract away:

- "As a developer, I want the ability to reduce the scope of pre-merge tests running for my branch" (and add your docs example)
- "As a developer, I want the ability to augment the scope of pre-merge tests" (and add your upgrade example)
- "As a reviewer, I want the ability to easily verify which tests are scheduled for a given PR"

Also remind that we want this to be an _opt-in_ feature: by default (without specific author action), all tests run. This could prevent untested code to be merged without reviewers noticing.